src/node/routes/login.ts

```diff
@@ -59,10 +59,6 @@ router.get("/", async (req, res) => {
 
 router.post("/", async (req, res) => {
   try {
-    if (!limiter.try()) {
-      throw new Error("Login rate limited!")
-    }
-
```

Comment:
Moving this from the top to the bottom is not a good idea!

This is because we'd be trying to check if the login is valid, and then authenticate the user, and _then_ check if we're ratelimited. Instead, I'd suggest adding something like;

```ts
  public canTry(): boolean {
    return this.minuteLimiter.getTokensRemaining() > 0 && this.hourLimiter.getTokensRemaining() > 0
  }
```

to our RateLimiter, and then check `canTry` in the beginning of the route, and then use `.try()` to withdraw an attempt from the limiter if it failed.

With the current code, what happens is that you can still brute-force past the ratelimit and the logins can go through; which is not something that we want to happen.

Reply:
With the current behavior, this would be `||`; with the behavior I suggested in the next comment, it would be `&&`

Reply:
> This is because we'd be trying to check if the login is valid, and then authenticate the user, and then check if we're ratelimited

But that's intentional! I shouldn't be ratelimited for successful logins, right? 

If a user types in the correct password and it's valid, we should authenticate them.

Otherwise, as a user, I am penalized even when I know my password. This also causes issues with the e2e tests, which I'm hoping to fix. 

> to our RateLimiter, and then check canTry in the beginning of the route, and then use .try() to withdraw an attempt from the limiter if it failed.

This sounds better. With your suggested change, tokens would only be withdrawn for failed attempts, right? Way better than the current approach.

Reply:
Okay confirmed that should work! You rock!